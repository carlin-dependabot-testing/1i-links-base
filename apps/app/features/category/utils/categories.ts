import { DEFAULT_CATEGORY } from '@/constants'
import { type Category } from '@/features/links/types/category.types'
import { getCategories } from '@/features/setup/utils/load-json-data'

/**
 * Returns a Set of category names in lowercase
 */
export const getCategoryNames = async (): Promise<Set<string>> => {
  const categories = await getCategories()

  return new Set([
    DEFAULT_CATEGORY,
    ...(categories?.map(cat => cat.name.toLowerCase()) || [])
  ])
}

/**
 * Returns all valid categories
 */
export const getValidCategories = async (): Promise<Category[]> => {
  const categories = await getCategories()
  return categories || []
}

/**
 * Validates if a category name exists
 */
export const isCategoryValid = async (
  categoryName: string
): Promise<boolean> => {
  const categoryNames = await getCategoryNames()
  return categoryNames.has(categoryName.toLowerCase())
}

/**
 * Returns a category by its slug
 */
export const getCategoryBySlug = async (slug: string) => {
  const categories = await getCategories()
  return categories?.find(category => category.name === slug) || null
}

// Synchronous versions with default values
export const getDefaultCategoryNames = (): Set<string> => {
  return new Set([DEFAULT_CATEGORY])
}

export const getDefaultCategories = (): Category[] => {
  return []
}
