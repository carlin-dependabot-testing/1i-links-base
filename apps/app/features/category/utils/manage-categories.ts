import fs from 'fs/promises'
import path from 'path'
import { z } from 'zod'

import { CategorySchema } from '@/features/links/schemas/category.schema'
import { type Category } from '@/features/links/types/category.types'

// Custom errors
export class CategoryError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'CategoryError'
  }
}

export const categoriesPath = path.join(
  process.cwd(),
  'public',
  'data',
  'categories.json'
)

/**
 * Retrieves all categories from the JSON file
 * @returns {Promise<Category[]>} Array of categories
 * @throws {CategoryError} If there's an error reading the file
 */
export async function getCategories(): Promise<Category[]> {
  try {
    const data = await fs.readFile(categoriesPath, 'utf8')

    if (data.trim().length === 0) {
      console.warn('The category file is empty')
      return []
    }

    try {
      const parsedData = JSON.parse(data)

      // Sort categories alphabetically by name (case-insensitive)
      return parsedData.sort((a: Category, b: Category) =>
        a.name.toLowerCase().localeCompare(b.name.toLowerCase())
      )
    } catch (parseError) {
      console.error('Error parsing JSON:', parseError)
      return []
    }
  } catch (error) {
    if (error instanceof z.ZodError) {
      console.error('Zod validation error:', error.errors)
      throw new CategoryError('Invalid data format in categories file')
    }
    console.error('Error reading category:', error)
    return []
  }
}

/**
 * Saves categories to the JSON file
 * @param {Category[]} category - Array of categories to save
 * @throws {CategoryError} If there's an error saving the file
 */
export async function saveCategories(category: Category[]): Promise<void> {
  try {
    const data = JSON.stringify(category, null, 2)
    await fs.writeFile(categoriesPath, data)
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to save categories')
  }
}

/**
 * Adds a new category to the storage
 * @param {Category} newCategory - Category to add
 * @returns {Promise<Category>} Added category
 * @throws {CategoryError} If the category is invalid or if there's an error saving
 */
export async function addCategory(category: Category): Promise<Category> {
  try {
    CategorySchema.parse(category)

    const categories = await getCategories()

    categories.push(category)
    await saveCategories(categories)

    return category
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to add category')
  }
}

export async function editCategory(
  oldName: string,
  newName: string,
  icon?: string,
  description?: string
): Promise<void> {
  try {
    if (!oldName?.trim() || !newName?.trim()) {
      throw new CategoryError('Invalid category name')
    }

    const categories = await getCategories()

    const categoryExists = categories.some(
      category => category.name === oldName
    )

    if (!categoryExists) {
      throw new CategoryError('Category not found')
    }

    const isDuplicate = categories.some(
      category => category.name === newName && category.name !== oldName
    )
    if (isDuplicate) {
      throw new CategoryError('A category with this name already exists')
    }

    const updatedCategories = categories.map((category: Category) => {
      if (category.name === oldName) {
        const updatedCategory = {
          ...category,
          name: newName,
          icon: icon || category.icon,
          description: description || category.description
        }

        // Validate the updated category
        CategorySchema.parse(updatedCategory)
        return updatedCategory
      }
      return category
    })

    await saveCategories(updatedCategories)
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to edit category')
  }
}

/**
 * Deletes a category from the storage
 * @param {string} name - Name of the category to delete
 * @throws {CategoryError} If the category name is invalid
 */
export async function deleteCategory(name: string): Promise<void> {
  try {
    if (typeof name !== 'string' || name.trim() === '') {
      throw new CategoryError('Invalid category name')
    }

    const categories = await getCategories()
    const updatedCategories = categories.filter(
      category => category.name !== name
    )
    await saveCategories(updatedCategories)
  } catch (error) {
    if (error instanceof CategoryError) {
      throw new CategoryError('Failed to delete category')
    }
    throw error
  }
}
