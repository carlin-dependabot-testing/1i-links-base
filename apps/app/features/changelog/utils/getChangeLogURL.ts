import appConfig from '@/config/app.config'

export function getChangeLogURL() {
  return `https://raw.githubusercontent.com/${appConfig.githubRepo}/refs/heads/main/CHANGELOG.md`
}
