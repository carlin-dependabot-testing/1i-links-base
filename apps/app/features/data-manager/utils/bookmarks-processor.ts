import { DEFAULT_CATEGORY, DEFAULT_COLOR } from '@/constants'
import { type LinksApp } from '@/features/links/types/link.types'

export async function processBookmarksFile(
  file: File
): Promise<Partial<LinksApp>[]> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()

    reader.onload = event => {
      try {
        const text = event.target?.result as string
        const parser = new DOMParser()
        const doc = parser.parseFromString(text, 'text/html')
        const bookmarks = Array.from(doc.getElementsByTagName('a'))

        const links = bookmarks.map(bookmark => ({
          name: bookmark.textContent || '',
          environments: {
            production: bookmark.href
          },
          category: DEFAULT_CATEGORY,
          color: DEFAULT_COLOR
        }))

        resolve(links.filter(link => link.name && link.environments.production))
      } catch (error) {
        reject(new Error('Failed to process bookmarks file'))
      }
    }

    reader.onerror = () => {
      reject(new Error('Failed to read bookmarks file'))
    }

    reader.readAsText(file)
  })
}
