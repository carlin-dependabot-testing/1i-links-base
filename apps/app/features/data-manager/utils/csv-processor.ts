import { type LinksApp } from '@/features/links/types/link.types'

export async function processCSVFile(file: File): Promise<Partial<LinksApp>[]> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()

    reader.onload = event => {
      try {
        const text = event.target?.result as string
        const lines = text.split('\n')
        const headers = lines[0]
          .toLowerCase()
          .split(',')
          .map(header => header.trim())

        const links = lines
          .slice(1)
          .filter(line => line.trim()) // Skip empty lines
          .map(line => {
            const values = line.split(',').map(value => {
              // Remove quotes and trim
              return value.trim().replace(/^"(.*)"$/, '$1')
            })

            const link: Partial<LinksApp> = {
              environments: { production: '' }
            }

            headers.forEach((header, index) => {
              const value = values[index]

              if (!value) return

              switch (header) {
                case 'name':
                  link.name = value
                  break
                case 'description':
                  link.description = value
                  break
                case 'category':
                  link.category = value
                  break
                case 'color':
                  link.color = value
                  break
                case 'tags':
                  link.tags = value
                    .split(';')
                    .map(tag => tag.trim())
                    .filter(Boolean)
                  break
                case 'production_url':
                  if (!link.environments) link.environments = { production: '' }
                  link.environments.production = value
                  break
                case 'staging_url':
                  if (!link.environments) link.environments = { production: '' }
                  link.environments.staging = value
                  break
                case 'integration_url':
                  if (!link.environments) link.environments = { production: '' }
                  link.environments.integration = value
                  break
              }
            })

            return link
          })
          .filter(link => {
            // Filter out invalid links
            return link.name && link.environments?.production
          })

        resolve(links)
      } catch (error) {
        console.error('CSV Processing Error:', error)
        reject(
          new Error('Failed to process CSV file. Please check the file format.')
        )
      }
    }

    reader.onerror = () => {
      reject(new Error('Failed to read CSV file'))
    }

    reader.readAsText(file)
  })
}
