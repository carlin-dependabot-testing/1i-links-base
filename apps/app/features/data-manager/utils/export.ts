import { type LinksApp } from '@/features/links/types/link.types'

import { type toast } from '@links-base/ui/hooks'

interface ExportOptions {
  links: Record<string, LinksApp[]>
  toast: typeof toast
}

export function exportAsHtml({ links, toast }: ExportOptions) {
  try {
    let html = '<!DOCTYPE NETSCAPE-Bookmark-file-1>\n'
    html +=
      '<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">\n'
    html += '<TITLE>Bookmarks</TITLE>\n'
    html += '<H1>Bookmarks</H1>\n'
    html += '<DL><p>\n'

    // Add category folders
    Object.entries(links)
      .sort()
      .forEach(([categoryName, categoryLinks]) => {
        html += `<DT><H3>${categoryName}</H3>\n<DL><p>\n`

        // Group links by base name for multi-environment links
        const groupedLinks = categoryLinks.reduce(
          (acc, link) => {
            if (
              link.environments &&
              Object.keys(link.environments).length > 1 &&
              link.name
            ) {
              if (!acc[link.name]) {
                acc[link.name] = []
              }
              acc[link.name].push(link)
            }
            return acc
          },
          {} as Record<string, LinksApp[]>
        )

        // Handle grouped environment links
        Object.entries(groupedLinks).forEach(([baseName, links]) => {
          if (!baseName) return

          html += `<DT><H3>${baseName}</H3>\n`
          html += '<DL><p>\n'

          const envLinks = links[0]?.environments
          if (envLinks) {
            Object.entries(envLinks)
              .sort(([a], [b]) => a.localeCompare(b))
              .forEach(([env, url]) => {
                if (url) {
                  const title = `${baseName} (${env})`
                  html += `<DT><A HREF="${url}">${title}</A>\n`
                }
              })
          }
          html += '</DL><p>\n'
        })

        // Handle regular links
        categoryLinks
          .filter(
            link =>
              !link.environments || Object.keys(link.environments).length <= 1
          )
          .forEach(link => {
            if (!link.name) return
            const url = link.environments?.production

            if (url) {
              html += `<DT><A HREF="${url}">${link.name}</A>\n`
            }
          })

        html += '</DL><p>\n'
      })

    html += '</DL>'

    const blob = new Blob([html], { type: 'text/html' })
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = 'bookmarks.html'
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(a)
    URL.revokeObjectURL(url)

    toast({
      title: 'Success',
      description: 'Bookmarks exported successfully as HTML'
    })
  } catch (error) {
    console.error('Export error:', error)
    toast({
      title: 'Error',
      description: 'Failed to export bookmarks',
      variant: 'destructive'
    })
  }
}
