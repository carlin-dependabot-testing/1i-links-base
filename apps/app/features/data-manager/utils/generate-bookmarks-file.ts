import { type LinksApp } from '@/features/links/types/link.types'

export const generateBookmarksFile = (
  webappsByCategory: Record<string, LinksApp[]>
) => {
  let bookmarksHtml = `
    <!DOCTYPE NETSCAPE-Bookmark-file-1>
    <!-- This is an automatically generated file.
         It will be read and overwritten.
         DO NOT EDIT! -->
    <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
    <TITLE>Bookmarks</TITLE>
    <H1>Bookmarks</H1>
    <DL><p>
  `

  Object.entries(webappsByCategory).forEach(([categoryName, links]) => {
    bookmarksHtml += `<DT><H3 ADD_DATE="${Math.floor(Date.now() / 1000)}" LAST_MODIFIED="${Math.floor(Date.now() / 1000)}">${categoryName}</H3><DL><p>`
    links.forEach(link => {
      bookmarksHtml += `<DT><A HREF="${link.environments?.production}" ADD_DATE="${Math.floor(Date.now() / 1000)}" ICON="${link.icon}">${link.name}</A></DT>`
    })
    bookmarksHtml += '</DL><p>'
  })

  bookmarksHtml += '</DL><p>'
  return bookmarksHtml
}
