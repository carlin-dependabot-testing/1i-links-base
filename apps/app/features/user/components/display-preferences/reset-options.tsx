'use client'

import { RefreshCw, RotateCcw } from 'lucide-react'
import { toast } from 'sonner'

type ResetOptionsProps = {
  onResetWarnings: () => void
  onResetDefaults: () => void
}

export const ResetOptions = ({
  onResetWarnings,
  onResetDefaults
}: ResetOptionsProps) => {
  return (
    <div className="flex flex-col gap-2">
      <div
        className="flex cursor-pointer items-center gap-2 text-sm hover:text-primary"
        onClick={() => {
          onResetWarnings()
          toast.success('All warning dialogs have been reset')
        }}
      >
        <RotateCcw className="h-4 w-4" />
        Reset Warning Dialogs
      </div>

      <div
        className="flex cursor-pointer items-center gap-2 text-sm hover:text-primary"
        onClick={() => {
          onResetDefaults()
          toast.success('Display preferences restored to defaults')
        }}
      >
        <RefreshCw className="h-4 w-4" />
        Reset to Default Settings
      </div>
    </div>
  )
}
