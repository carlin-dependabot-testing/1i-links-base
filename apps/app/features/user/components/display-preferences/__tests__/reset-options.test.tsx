import { toast } from 'sonner'

import { fireEvent, render, screen } from '@/utils/test-utils'

import { ResetOptions } from '../reset-options'

// Mock sonner toast
jest.mock('sonner')

const renderComponent = (
  props: Partial<Parameters<typeof ResetOptions>[0]> = {}
) => {
  const defaultProps = {
    onResetWarnings: jest.fn(),
    onResetDefaults: jest.fn()
  }

  return render(<ResetOptions {...defaultProps} {...props} />)
}

describe('ResetOptions', () => {
  it('ResetOptions should render correctly', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('should display both reset options', () => {
    renderComponent()

    expect(screen.getByText('Reset Warning Dialogs')).toBeInTheDocument()
    expect(screen.getByText('Reset to Default Settings')).toBeInTheDocument()
  })

  it('should call onResetWarnings and show success toast when clicking reset warnings', () => {
    const onResetWarnings = jest.fn()
    renderComponent({ onResetWarnings })

    fireEvent.click(screen.getByText('Reset Warning Dialogs'))

    expect(onResetWarnings).toHaveBeenCalledTimes(1)
    expect(toast.success).toHaveBeenCalledWith(
      'All warning dialogs have been reset'
    )
  })

  it('should call onResetDefaults and show success toast when clicking reset defaults', () => {
    const onResetDefaults = jest.fn()
    renderComponent({ onResetDefaults })

    fireEvent.click(screen.getByText('Reset to Default Settings'))

    expect(onResetDefaults).toHaveBeenCalledTimes(1)
    expect(toast.success).toHaveBeenCalledWith(
      'Display preferences restored to defaults'
    )
  })
})
