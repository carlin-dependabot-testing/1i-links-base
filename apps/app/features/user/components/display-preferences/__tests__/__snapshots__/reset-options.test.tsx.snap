// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetOptions ResetOptions should render correctly 1`] = `
<div>
  <div
    class="group/sidebar-wrapper flex min-h-svh w-full has-[[data-variant=inset]]:bg-sidebar"
    style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
  >
    <script
      nonce=""
    >
      ((e,s,n,m,o,l,u,h)=&gt;{let c=document.documentElement,T=["light","dark"];function p(d){(Array.isArray(e)?e:[e]).forEach(g=&gt;{let k=g==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(c.classList.remove(...S),c.classList.add(d)):c.setAttribute(g,d)}),C(d)}function C(d){h&&T.includes(d)&&(c.style.colorScheme=d)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let d=localStorage.getItem(s)||n,g=u&&d==="system"?a():d;p(g)}catch(d){}})("class","theme","system",null,["light","dark"],null,true,true)
    </script>
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="flex cursor-pointer items-center gap-2 text-sm hover:text-primary"
      >
        <svg
          class="lucide lucide-rotate-ccw h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 1 0 9-9 9.75 9.75 0 0 0-6.74 2.74L3 8"
          />
          <path
            d="M3 3v5h5"
          />
        </svg>
        Reset Warning Dialogs
      </div>
      <div
        class="flex cursor-pointer items-center gap-2 text-sm hover:text-primary"
      >
        <svg
          class="lucide lucide-refresh-cw h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
          />
          <path
            d="M21 3v5h-5"
          />
          <path
            d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
          />
          <path
            d="M8 16H3v5"
          />
        </svg>
        Reset to Default Settings
      </div>
    </div>
    <div
      class="fixed bottom-4 right-4 z-50 flex items-center gap-2 rounded-lg bg-background/80 p-2 shadow-sm backdrop-blur-sm"
    />
    <div
      aria-label="Notifications (F8)"
      role="region"
      style="pointer-events: none;"
      tabindex="-1"
    >
      <ol
        class="fixed top-0 z-[100] flex max-h-screen w-full flex-col-reverse p-4 sm:bottom-0 sm:right-0 sm:top-auto sm:flex-col md:max-w-[420px]"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
